lms/djangoapps/grades/models.py

```diff
@@ -690,6 +676,59 @@ def get_override(cls, user_id, usage_key):
         except PersistentSubsectionGradeOverride.DoesNotExist:
             pass
 
+    @classmethod
+    def update_or_create_override(
+        cls, requesting_user, subsection_grade_model, feature=None, action=None, **override_data
+    ):
+        """
+        Creates or updates an override object for the given PersistentSubsectionGrade.
+        Args:
+            requesting_user: The user that is creating the override (so we can record this action in
+            a PersistentSubsectionGradeOverrideHistory record).
+            subsection_grade_model: The PersistentSubsectionGrade object associated with this override.
+            override_data: The parameters of score values used to create the override record.
+        """
+        override, _ = PersistentSubsectionGradeOverride.objects.update_or_create(
+            grade=subsection_grade_model,
+            defaults=cls._prepare_override_params(subsection_grade_model, override_data),
+        )
+
+        feature = feature or PersistentSubsectionGradeOverrideHistory.GRADEBOOK
+        action = action or PersistentSubsectionGradeOverrideHistory.CREATE_OR_UPDATE
+
+        PersistentSubsectionGradeOverrideHistory.objects.create(
+            override_id=override.id,
+            user=requesting_user,
+            feature=feature,
+            action=action,
+        )
+        return override
+
+    @staticmethod
+    def _prepare_override_params(subsection_grade_model, override_data):
+        """
+        Helper method to strip any grade override field names that won't work
+        as defaults when calling PersistentSubsectionGradeOverride.update_or_create(),
+        and to use default values from the associated PersistentSubsectionGrade
+        for any override fields that are not specified.
+        """
+        allowed_fields_and_defaults = {
+            'earned_all_override': 'earned_all',
+            'possible_all_override': 'possible_all',
+            'earned_graded_override': 'earned_graded',
+            'possible_graded_override': 'possible_graded',
+        }
+        cleaned_data = {}
+        for field in override_data.keys():
+            if field in allowed_fields_and_defaults:
+                cleaned_data[field] = override_data[field]
+
+        for override_field, default_field in allowed_fields_and_defaults.items():
+            if override_field not in cleaned_data:
+                cleaned_data[override_field] = getattr(subsection_grade_model, default_field)
```

Comment:
Can L722-728 be simplified to the following?

```
for override_field_name, field_name in allowed_fields_and_defaults.items():
    cleaned_data[override_field_name] = getattr(
        override_data,
        override_field_name, 
        getattr(subsection_grade_model, field_name),
    )
```

Reply:
Yes, something like that should work.